sql/src/main/java/io/crate/analyze/AlterTableRerouteAnalyzer.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.analyze;
+
+import com.google.common.collect.ImmutableSet;
+import io.crate.analyze.expressions.ExpressionToNumberVisitor;
+import io.crate.analyze.expressions.ExpressionToStringVisitor;
+import io.crate.data.Row;
+import io.crate.metadata.PartitionName;
+import io.crate.metadata.Schemas;
+import io.crate.metadata.TableIdent;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.metadata.table.Operation;
+import io.crate.metadata.table.TableInfo;
+import io.crate.sql.tree.AlterTableReroute;
+import io.crate.sql.tree.AstVisitor;
+import io.crate.sql.tree.Expression;
+import io.crate.sql.tree.RerouteAllocateReplicaShard;
+import io.crate.sql.tree.RerouteCancelShard;
+import io.crate.sql.tree.RerouteMoveShard;
+import io.crate.sql.tree.RerouteRetryFailed;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+
+public class AlterTableRerouteAnalyzer {
+
+    private final Schemas schemas;
+    private static final RerouteOptionVisitor REROUTE_OPTION_VISITOR = new RerouteOptionVisitor();
+
+    AlterTableRerouteAnalyzer(Schemas schemas) {
+        this.schemas = schemas;
+    }
+
+    public AnalyzedStatement analyze(AlterTableReroute node, Analysis context) {
+        TableInfo tableInfo = schemas.getTableInfo(
+            TableIdent.of(node.table(), context.sessionContext().defaultSchema()),
+            Operation.ALTER);
+        Row parameters = context.parameterContext().parameters();
+        PartitionName partitionName = AlterTableAnalyzer.createPartitionName(node.table().partitionProperties(),
+            (DocTableInfo) tableInfo, parameters);
+        return REROUTE_OPTION_VISITOR.process(node.rerouteOption(), new Context(tableInfo, partitionName, parameters));
+    }
+
+    private class Context {
+
+        final TableInfo tableInfo;
+        final PartitionName partitionName;
+        final Row params;
+
+        private Context(TableInfo tableInfo, PartitionName partitionName, Row params) {
+            this.tableInfo = tableInfo;
+            this.partitionName = partitionName;
+            this.params = params;
+        }
+
+    }
+
+    private static class RerouteOptionVisitor extends AstVisitor<RerouteAnalyzedStatement, Context> {
+
+        private static final Set<String> ALLOWED_CANCEL_OPTIONS = ImmutableSet.of("allow_primary");
```

Comment:
as it contains only 1 entry, it would make sense to avoid using a set, but a single string instead